proposals/text/0000-shaper-in-prompt-template.md

```diff
@@ -0,0 +1,153 @@
+
+
+# Detailed design
+
+PromptTemplate gets two new attributes: `input_shapers` and `output_shapers`. These are lists of Shaper objects that are applied to the input and output of the prompt respectively. The `input_shapers` are applied to the input of the prompt, the `output_shapers` are applied to the output of the prompt:
+PromptNode calls `PromptTemplate.prepare` before exectuting the prompt. `PromptTemplate.prepare` makes all `input_shapers` run on the `invocation_context`.
+PromptNode invokes the prompt on the prepared `invocation_context`.
+PromptNode calls `PromptTemplate.post_process` after executing the prompt. `PromptTemplate.post_process` makes all `output_shapers` run on the `invocation_context`.
+
+Note, that `Shapers` are still usable in Pipelines as before.
+
+# Drawbacks
+
+Look at the feature from the other side: what are the reasons why we should _not_ work on it? Consider the following:
+
+- What's the implementation cost, both in terms of code size and complexity? A good day
+- Can the solution you're proposing be implemented as a separate package, outside of Haystack? No
+- Does it teach people more about Haystack? No, but it makes it easier to use especially for beginners.
+- How does this feature integrate with other existing and planned features? It doesn't change any existing features and should nicely integrate with agents.
+- What's the cost of migrating existing Haystack pipelines (is it a breaking change?)? None
+
+It also fosters a bit the nesting of components in Haystack. Although the whole PromptNode ecosystem already does this (e.g. via PromptModel, PromptTemplate being used by PromptNode), it's still a bit of a new concept. However, I think it's a good one and it's not too hard to understand.
+
+We still don't have access to PromptNode, PromptModel or the invocation layer inside of PromptTemplates. If we want Shapers to access fundamental parts of them (e.g. the tokenizer), we would need to pass them to the Shapers. This would make the whole system more complex, but it would be possible.
+
+# Alternatives
+
+Subclassing specialiced PromptNodes like QuestionAnsweringPromptNode, which would have the shapers already defined. This would make it easier to use, but it would be harder to iterate quickly on prompts, be less flexible and sharing is difficult. Also we get
+
+The same is true for subclassing PromptTemplate like QuestionAnsweringPromptTemplate. This would make it easier to use, but it would be harder to iterate quickly on prompts, be less flexible and sharing is difficult.
+
+# Adoption strategy
+
+No need for migration, just adjusting the docs and the promptnode tutorial (if it exists already).
+
+# How we teach this
+
+We should show how:
+- predefined PromptTemplates can be used
+- predefined PromptTemplates can be customized
+- custom PromptTemplates can be created
+
+# Unresolved questions
+
+- how to deal with usage of PromptTemplate outside of PromptNode? (e.g. in OpenAIAnswerGenerator)
```

Comment:
We recently added support for `PromptTemplate` in `OpenAIAnswerGenerator`. It's not full support because we are missing the features that are described here. Basically, the addition of the input_shapers and output_shapers would allow us to fully use the `PromptTemplate` without having hard-coded delimiter values within `OpenAIAnswerGenerator`.

Also ultimately, I think `OpenAIAnswerGenerator` should just become wrapper around a `PromptNode`. However, this would not be possible until we manage the "intelligent" truncation I explained in this [comment](https://github.com/deepset-ai/haystack/pull/4172/files#r1113096039).